### PR TITLE
Restore targeting and other fixes

### PR DESCRIPTION
**Pull request by squintfox**
@olofvndrhr  Hey, I restored some of the original upstream functionality that allows records to be imported into NetBox.  I also fixed a couple of problems that existed in the original version.

### CODE REVIEW

**Comment by squintfox** [comment]
I think populate() is generally supposed to return True/False if the zone exists or doesn't, but there's a few pieces to that it seems.  For now I took it out.

**Comment by olofvndrhr** [comment]
I just checked the source base class and you're right.
https://github.com/octodns/octodns/blob/9f808274f57cff343af7e4175b80a6f46815ca41/octodns/source/base.py#L43-L44

The weird thing is, when i restuctured the code i checked the original sources (yaml and envvar) and they both don't return anything. E.g:
https://github.com/octodns/octodns/blob/9f808274f57cff343af7e4175b80a6f46815ca41/octodns/source/envvar.py#L91

But the providers do. E.g powerdns:
https://github.com/octodns/octodns-powerdns/blob/80af99402747b63f07308c14f821edd4ea4dfb5f/octodns_powerdns/__init__.py#L485C16-L485C22


Thats a bit weird. But i think i should also add it.
When i forked the original repo i just wanted to make the "views" filter work and add some minor functionality like the "@" symbol as the zone name, but i didnt really check if everything was implemented correctly.
But i should probably take a good look and fix/add missing things.

About your changes: i have to test them locally as i currently only use netbox as a source. But after that i will merge it and update the README etc.

File: src/octodns_netbox_dns/__init__.py [hunk 1]
```diff
@@ -193,7 +196,7 @@ def _format_rdata(self, rdata: dns.rdata.Rdata, raw_value: str) -> str | dict[st
                 }
 
             case "SPF" | "TXT":
-                value = raw_value.replace(";", r"\;")
+                value = raw_value.replace(";", "\\;")
```
Comment (olofvndrhr) [inline src/octodns_netbox_dns/__init__.py:199]: I would like to keep the raw string as its clearer what the replacement exactly is.

Comment (squintfox) [reply src/octodns_netbox_dns/__init__.py:199]: The reason I did that is because in lines 325 and 402, I needed to use a single backslash and ```r"\"``` isn't valid, so I used ```"\\"```.  Since I had that one in that syntax, I made all the others the same.  It's your call if you want to keep just that one as raw, make everything else except the singular ```"\\"``` raw or leave it the way I put it.

Unless you know a different syntax that gives you something equivalent to ```r"\"```.

**Comment by squintfox** [comment]
It might just be that the "source only" ones, like YAML don't return because it only matters in a target scenario?  I'm not sure why existence matters, but maybe it's just to know whether to create/not create on the target side.

File: src/octodns_netbox_dns/__init__.py (hunk 1 above)
Comment (olofvndrhr) [reply src/octodns_netbox_dns/__init__.py:199]: Ah i get it. Then its ok and we keep it all the same (your approach)

**Comment by olofvndrhr** [comment]
> It might just be that the "source only" ones, like YAML don't return because it only matters in a target scenario? I'm not sure why existence matters, but maybe it's just to know whether to create/not create on the target side.

~Yes, probably because even the providers can't create a zone, it has to be present to be updated.~

But i find the overall docs for octodns not very "good" to create custom providers/sources.
You have to search in different places which function you have to overwrite with the subclass and then also check the official providers for reference.

**Comment by squintfox** [comment]
The Route53 provider can create zones:
https://github.com/octodns/octodns-route53/blob/9dff89bf677a546af1beb7301d957222e1f0e44e/octodns_route53/provider.py#L855

The Azure one can as well:
https://github.com/octodns/octodns-azure/blob/d7a41976ebad42656749a560199060707f0de6f2/octodns_azure/__init__.py#L685

However, it's probably impractical in NetBox DNS because you'd have to supply all the required SOA values at zone creation, plus you'd have to make it able to create name server objects (though this part is doable).

I agree that the "documentation" isn't really aggregated or well-explained.  I've basically just been banging my head against it as you are as well.

**Comment by olofvndrhr** [comment]
Hello @squintfox
I refactored and simplified some of your changes in the dev branch.
I also added a filter for creating records which ignores `NS`,`SOA` and `PTR` records as they give errors if they try to get created.
I tested everything locally and it works. Can you also do some testing?
After that i will merge it to main and create a new release.

Also: should the Class be renamed to `NetBoxDNSProvider` as its not only a source anymore?

Best regards

**Comment by squintfox** [comment]
@olofvndrhr At a glance it all looks great.  I'll pull it today and try it out.

Canonically it should definitely be NetBoxDNSProvider.  I actually made that change originally but I didn't submit it because I didn't want to introduce a breaking change without discussion.  I'm all for changing it, but I don't know if there's any way it can retain backward compatibility as well.

**Comment by olofvndrhr** [comment]
To be honest, i don't think it matters if its breaking as its a relatively new fork and probably only us two currently using it.
We probably could split the functionality between source and provider (and also the classes). But i think that's something for the future as its not really necessary.

**Comment by squintfox** [comment]
Then I'm all for the rename.

**Comment by squintfox** [comment]
I pulled the dev branch and had no issues running through all the things I've been working on the last few weeks.  Looks good to me.

**Comment by olofvndrhr** [comment]
Perfect, thanks for the testing!
I will write some more automated tests tomorrow and then merge+release.

**Comment by squintfox** [comment]
@olofvndrhr Something seems to have broken between dev and the pypi release.  I pulled the newest and it's crashing any time it reads an NS record out of NetBox.  Trying to track it down.

**Comment by squintfox** [comment]
@olofvndrhr Found it.  This line has a typo.

https://github.com/olofvndrhr/octodns-netbox-dns/blob/6b06cada38751648b9365fd27b850aa2b6d250a8/src/octodns_netbox_dns/__init__.py#L256

Should be:

```rcd_rdata = self._format_rdata(rcd_type, rcd_value)```

Also, this doesn't really work any more:

https://github.com/olofvndrhr/octodns-netbox-dns/blob/6b06cada38751648b9365fd27b850aa2b6d250a8/src/octodns_netbox_dns/__init__.py#L253

It just shows blank list every time.  I'm not exactly sure how it should be.

**Comment by olofvndrhr** [comment]
@squintfox
Yes i checked it too now. Such a stupid typo.
I will fix and re-release it.

The record data works for me:
image

**Comment by squintfox** [comment]
@olofvndrhr Yeah I just meant that it always just says values=[].  Maybe I'm just not putting the pieces together.

Also this doesn't work fully either:
```value=re.sub(r"\\*;", ";", record),```

Trying to track down the specifics.  This value in a TXT record has too many slashes:  ```t=y\;o=~\;```

**Comment by olofvndrhr** [comment]
I pushed some changes which fixes the issues.
This time i tested the syncs in both ways and found no issues.

Your issue with the TXT record i didnt find. I also added some tests with your example, but they all work.

**Comment by squintfox** [comment]
@olofvndrhr Let me test the new version and if it's still not working for me I'll make a more detailed writeup.  I'm not recalling all the backslash and semicolon stuff I figured out like a month ago.  There are multiple things at play in both how NetBox API outputs backslashes and how octodns handles backslash/semicolon internally.

**Comment by squintfox** [comment]
@olofvndrhr OK, I finally figured it out.  Originally the code used to say this, which worked fine:
```
case "SPF" | "TXT":
	value = raw_value.replace(";", r"\;")
```
This is because octo requires that all semicolons be escaped internally and asks the providers to do it on the way in.  Otherwise, you get something like this:
```
octodns.record.exception.ValidationError: Invalid record "test.com.", system/dns\test.com.yaml, line 20, column 5
  - unescaped ; in "k=rsa; p=Msfkjdshfyh789hsdckijashdkjkhdsbfs8ysd8fhaoiudhfB;"
```

However, now it says this:
https://github.com/olofvndrhr/octodns-netbox-dns/blob/a8ecfab93096cd3201efcb06a65bc86a2444611a/src/octodns_netbox_dns/__init__.py#L105

This means that using the earlier test case of ```t=y\;o=~\;```, it currently evaluates to ```t=y\;o=~\;```.  However, because of the internal escaping, it needs to be represented as ```t=y\\;o=~\\;```.

If you use any other provider and send it to YAML, it shows this as well.  The YAML shows all source semicolons as ```\;```.  If you try to load any YAML with only a semicolon, it'll raise an exception like above.

Long story short, I think this fixes it (or so does the original code you had before all these updates):
```value = re.sub(";", "\\;", value)```

Instead of replacing "all the backslashes or none plus semicolon" with "backslash semicolon", you need to replace every semicolon regardless of it's leading slashes with ```\;```.  octodns figures it out internally and so do all the other providers.

**Comment by olofvndrhr** [comment]
Hello @squintfox

> This is because octo requires that all semicolons be escaped internally and asks the providers to do it on the way in. Otherwise, you get something like this:

Yes, the error could also be "ignored" if you set `lenient` to true in the config.
But i changed the old replace as to when there are multiple backslashes (escapes) for semicolons which normally are not needed.

Are you sure that your record needs the semicolon escaped in netbox? Normally its not needed.
Normally this should work:
`t=y; o=~;`
The same is for example valid for DMARC records:
`v=DMARC1; p=none; ...`

**Comment by olofvndrhr** [comment]
I checked some other providers and they all do simple escaping and un-escaping.
So i will just follow that and revert my changes.

**Comment by squintfox** [comment]
@olofvndrhr Right, but ```t=y; o=~;``` is not the text for the record.  The correct text on an nslookup is ```t=y\; o=~\;```.

I also realized the other regex is problematic too.  Given a record of ```12;3ndkj\jddsjkj\\\\\\\\;```, you just lose all the slashes with this:
https://github.com/olofvndrhr/octodns-netbox-dns/blob/a8ecfab93096cd3201efcb06a65bc86a2444611a/src/octodns_netbox_dns/__init__.py#L107

Output: ```12;3ndkj\jddsjkj;```

The original code using replace() every ```;```with ```\;``` and on create replace() ```\;``` with ```;``` is, I think, correct.

**Comment by olofvndrhr** [comment]
@squintfox 
I think you are right.
I reverted to the old method (but in a separate function to see it more clearly what is done where).

On your old implementation for creating records you also replaced all double escapes, i think this would cause the same issue you explained above:
```python
value=record.replace("\\\\", "\\").replace("\\;", ";"),
```

**Comment by squintfox** [comment]
@olofvndrhr You actually fixed that in another way, possibly inadvertently.  NetBox's API (if you look at it in the GUI) returns 2 backslashes for every slash.  In the original code, that's how pynetbox was returning it.  But with your refactor, calling Record.value does it for us automatically, so it's just not needed anymore.

**Comment by olofvndrhr** [comment]
@squintfox 
I really hate this semicolon escaping stuff.
But i used your initial solution again and it seems to work best.
As i only use it as a source, my testing is and will be way less accurate as yours. So i trust you here.

Hope everything works now...finally.

**Comment by squintfox** [comment]
@olofvndrhr Lol, I have the benefit of an org with 10+ years of DNS mismanagement, so I've got all sorts of crazy records in another provider...

Let me pull and re-test.  Thanks for all the great work.

**Comment by olofvndrhr** [comment]
~~I also tested another solution which works with my tests, but of couse i only have about 5 TXT records which i test with.~~
```python
value.replace("\\\\;", ";")
```

Im not really sure why, but the request to create a new record always double escapes the semicolon even if its just normally escaped in the yaml:
```yml
value: v=DKIM1\; k=rsa\; p=MIIBI
```
the request:
```bash
| Create | _smtp._tls | TXT | 3600 | v=TLSRPTv1\; rua=mailto:tlsrpt@example.com | config |
-->
2024-02-29T23:28:22  [140243665496960] DEBUG NetboxDNSSource[netbox] ADD TXT _smtp._tls v=TLSRPTv1\\; rua=mailto:tlsrpt@example.com
```
~~So im not sure if its better to only strip the double escaped semicolons to not mess with any other record backslashes.~~

Also, we should probably only unescape `TXT` and `SPF` record on the `_apply`. As we also only do it in `populate` for these
> already done now

**Comment by squintfox** [comment]
@olofvndrhr Let me regroup tomorrow.  I've been testing more widely and it seems that between all the DNS providers everywhere double slashes are handled differently (the same TXT record with double backslashes returned from different internet DNS servers is not always the same).  I'm trying to figure out what's right and if not at least figure out what the providers that I have access to handle in a way that makes sense to me.

We're close.

**Comment by olofvndrhr** [comment]
@squintfox 
In my newest commit on dev i fixed an issue of the changeset.
Before the unescaping was done, the changes were compared to the current records, so it was not able to correctly guess all changes for txt records.
For example, if you changed a value in your txt record, it would not update it, but create a new one as the old one which it wanted to delete didnt exist for octodns (not escaped)

**Comment by squintfox** [comment]
@olofvndrhr So I did something not real smart and didn't carefully read your note before I picked this up again.  I didn't pull dev, I pulled main and managed to re-figure out what you already did.  The 2 remaining issues were:

1) Needed to remove the double backslashes from the changeset records.  I'm not 100% where that comes from but it must be something inside octodns.
2) Re-querying the NetBox records was giving unescaped values in the apply and they weren't matching to the changeset.

All that said, I think you've solved it.  I tested the dev branch and can't find any further issues.

And to boot, I found an issue in the UltraDNS octo provider.  That double backslash handling wasn't working correctly either.  I'm going to submit a PR over there.  That slowed me down yesterday a pretty good bit until I realized there were 2 sides to the problem.

**Comment by olofvndrhr** [comment]
@squintfox 
I probably found out where the backslashes get escaped again.
If i "debug" what is done with the changeset it becomes clear:
image

image
image
The `repr` call from the original implementation doesn't remove the backslashes from the octodns changeset. It also keeps the quotes for it, thats why there is this weird slicing of the string.

The fstring/`__str__` representation seems to keep everything original. So im not sure if we should change it to that to not use this weird double backslash unescaping we do, and return to a normal `.replace(r"\;", ";")`.

What do your think?

**Comment by squintfox** [comment]
@olofvndrhr f-string certainly seems more straightforward to me, but it makes me wonder why someone wrote that in the first place (seems like a weird choice without a good reason).  What is that implementation covering up that we don't know?

I guess my vote is, let's change it to f-strings and do a full re-test and see what happens.

**Comment by olofvndrhr** [comment]
@squintfox im note sure why its that way. But im also not sure why the original implementation chose to use `repr` instead of `str`.
Maybe there is a reason and we will find out after you test again ;)
My local tests work fine with the new changes. I will push them to dev shortly.

**Comment by squintfox** [comment]
@olofvndrhr Something definitely went wrong.  Something like, octodns expects records that start with a char to be without single quotes and if they start with an int, they need single quotes. Trying to track down the specifics.  As it stands, A records are OK (ints) and MX (char) make things crash.

**Comment by olofvndrhr** [comment]
@squintfox i reverted the changes from `repr` to `str`.
As it worked on all your tests with this solution and i don't really want to search everywhere in the octodns code where the problem lies, i think we should just keep it like that.

**Comment by olofvndrhr** [comment]
Also, could you add some test cases with your "special" records? (as a PR)
I don't really have a big arsenal of different records which could produce errors (double escaped, backslashes, other weird stuff).
Would be really appreciated, thanks!

**Comment by squintfox** [comment]
@olofvndrhr Well I figured out why repr() is in use:
https://github.com/octodns/octodns/blob/3acfd174b61ff5352294a0fbb5f9fb706d76bf14/octodns/record/dynamic.py#L425-L441

I think we need to stick with the previous way.  repr().replace('\\\\', '\\')

**Comment by squintfox** [comment]
@olofvndrhr Absolutely.  There's something else going on with "you have 2 records with the same name in NetBox" (totally valid) and the "modify" fires for the record you did not change in the source as well as the "delete" for the one you did (I think that means you overwrite the TTL from the actually changed record to all records with that name).  I'm tapped out for today, but I'll get back to this over the weekend or Monday.  Thanks again.

**Comment by olofvndrhr** [comment]
@squintfox 
Ah i see now why the `repr` exists. But unfortunately until this "hack" is fixed in octodns we need to live with it.
The current state of the dev branch should now work with everything you tested before -> use repr, escaped backslashes etc.
If everything works well for you, i will release a new version.

**Comment by squintfox** [comment]
@olofvndrhr I officially give the squintfox stamp of approval to the dev branch.  Turns out the "multiple records vs TTL situation" is more of a fix than a bug.  Netbox DNS lets you do something that's incorrect (have 2 records with the same name and different TTL values).  So, the provider takes the opportunity to update the TTL on all records of that name if there's any update made at all.  It's not 100% perfect because the Provider does only populate one TTL when 2 exist in NetBox, but I don't think there's anything you or I can do about that.  I may open a discussion with the plugin creator and see if there's a solution on that end.

**Comment by olofvndrhr** [comment]
Thanks for all your help!
New release was created and published.

**Comment by squintfox** [comment]
@olofvndrhr I reached the point of "I just cannot figure this out myself".  Let's see what the core devs think:
https://github.com/octodns/octodns/issues/1145
